web/src/banners.ts

```diff
@@ -0,0 +1,50 @@
+import type Handlebars from "handlebars/runtime.js";
+import $ from "jquery";
+
+import render_banner from "../templates/components/banner.hbs";
+
+type ComponentIntent = "neutral" | "brand" | "info" | "success" | "warning" | "danger";
+
+type ActionButton = {
+    type: "primary" | "quiet" | "borderless";
+    intent?: ComponentIntent;
+    label: string;
+    icon?: string;
+    custom_classes?: string;
+};
+
+export type Banner = {
+    intent: ComponentIntent;
+    label: string | Handlebars.SafeString;
+    buttons: ActionButton[];
+    close_button: boolean;
+    custom_classes?: string;
+};
+
+export type AlertBanner = Banner & {
+    process: string;
+};
+
+export function open(banner: Banner | AlertBanner, $banner_container: JQuery): void {
+    const banner_html = render_banner(banner);
+    $banner_container.html(banner_html);
+    $(window).trigger("resize");
+}
+
+export function append(banner: Banner | AlertBanner, $banner_container: JQuery): void {
+    const $banner_html = render_banner(banner);
+    $banner_container.append($banner_html);
+    $(window).trigger("resize");
+}
+
+export function close($banner: JQuery): void {
+    $banner.remove();
+    $(window).trigger("resize");
+}
+
+export function initialize(): void {
+    $("body").on("click", ".banner .banner_close_action", function (this: HTMLElement) {
+        const $banner = $(this).closest(".banner");
+        close($banner);
+    });
```

Comment:
Should this be using `preventDefault` or `stopPropagation`? I'm not sure it matters; the main risk I see is the default behavior of moving focus to the X button.

Reply:
Added. Made sense to add `e.preventDefault()` and `e.stopPropagation()` here since the event handler is registered on the body.